modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/MultiplexerTokenFilterFactory.java

```diff
@@ -72,23 +77,34 @@ public TokenStream create(TokenStream tokenStream) {
     }
 
     @Override
-    public void setReferences(Map<String, TokenFilterFactory> factories) {
-        filters = new ArrayList<>();
-        if (preserveOriginal) {
-            filters.add(IDENTITY_FACTORY);
-        }
-        for (String filter : filterNames) {
-            String[] parts = Strings.tokenizeToStringArray(filter, ",");
-            if (parts.length == 1) {
-                filters.add(resolveFilterFactory(factories, parts[0]));
-            } else {
-                List<TokenFilterFactory> chain = new ArrayList<>();
-                for (String subfilter : parts) {
-                    chain.add(resolveFilterFactory(factories, subfilter));
+    public TokenFilterFactory getChainAwareTokenFilterFactory(TokenizerFactory tokenizer, List<CharFilterFactory> charFilters,
+                                                              List<TokenFilterFactory> previousTokenFilters,
+                                                              Function<String, TokenFilterFactory> allFilters) {
+        if (filters == null) {
```

Comment:
should we worry about concurrency? Caching makes me a little uncomfortable due to the fact that it implies we expect that this will always be called with the same arguments, should we change the API or try to detect misusage?